src/data/energy.ts

```diff
@@ -915,25 +933,125 @@ export const computeConsumptionData = memoizeOne(
-      }
+  const outData: EnergyConsumptionData = {
+    used_total: {},
+    grid_to_battery: {},
+    battery_to_grid: {},
+    used_solar: {},
+    used_grid: {},
+    used_battery: {},
+    total: {
+      used_total: 0,
+      grid_to_battery: 0,
+      battery_to_grid: 0,
+      used_solar: 0,
+      used_grid: 0,
+      used_battery: 0,
+    },
+  };
+
+  data.timestamps.forEach((t) => {
+    const used_total =
+      (data.from_grid?.[t] || 0) +
+      (data.solar?.[t] || 0) +
+      (data.from_battery?.[t] || 0) -
+      (data.to_grid?.[t] || 0) -
+      (data.to_battery?.[t] || 0);
+
+    outData.used_total[t] = used_total;
+    outData.total.used_total += used_total;
+    const {
+      grid_to_battery,
+      battery_to_grid,
+      used_solar,
+      used_grid,
+      used_battery,
+    } = computeConsumptionSingle({
+      from_grid: data.from_grid && (data.from_grid[t] ?? 0),
+      to_grid: data.to_grid && (data.to_grid[t] ?? 0),
+      solar: data.solar && (data.solar[t] ?? 0),
+      to_battery: data.to_battery && (data.to_battery[t] ?? 0),
+      from_battery: data.from_battery && (data.from_battery[t] ?? 0),
     });
+
+    outData.grid_to_battery[t] = grid_to_battery;
+    outData.total.grid_to_battery += grid_to_battery;
+    outData.battery_to_grid![t] = battery_to_grid;
+    outData.total.battery_to_grid += battery_to_grid;
+    outData.used_battery![t] = used_battery;
+    outData.total.used_battery += used_battery;
+    outData.used_grid![t] = used_grid;
+    outData.total.used_grid += used_grid;
+    outData.used_solar![t] = used_solar;
+    outData.total.used_solar += used_solar;
   });
 
   return outData;
 };
 
+export const computeConsumptionSingle = (data: {
+  from_grid: number | undefined;
+  to_grid: number | undefined;
+  solar: number | undefined;
+  to_battery: number | undefined;
+  from_battery: number | undefined;
+}): {
+  grid_to_battery: number;
+  battery_to_grid: number;
+  used_solar: number;
+  used_grid: number;
+  used_battery: number;
+} => {
+  const to_grid = data.to_grid;
+  const to_battery = data.to_battery;
+  const solar = data.solar;
+  const from_grid = data.from_grid;
+  const from_battery = data.from_battery;
+
+  let used_solar = 0;
+  let grid_to_battery = 0;
+  let battery_to_grid = 0;
+  let used_battery = 0;
+  let used_grid = 0;
+  if ((to_grid != null || to_battery != null) && solar != null) {
+    used_solar = (solar || 0) - (to_grid || 0) - (to_battery || 0);
+    if (used_solar < 0) {
+      if (to_battery != null) {
+        grid_to_battery = used_solar * -1;
+        if (grid_to_battery > (from_grid || 0)) {
+          battery_to_grid = grid_to_battery - (from_grid || 0);
+          grid_to_battery = from_grid || 0;
+        }
+      }
+      used_solar = 0;
+    }
+  }
+
+  if (from_battery != null) {
+    if (to_grid != null) {
+      used_battery = (from_battery || 0) - (battery_to_grid || 0);
+    } else {
+      used_battery = from_battery;
+    }
```

Comment:
```suggestion
    used_battery = (from_battery || 0) - (battery_to_grid || 0);
```
I think the `if` is redundant here as `battery_to_grid` will be 0 in the else anyway but having it more explicit like this is ok too